### PR TITLE
App/Toponaming: Add MappedElement

### LINKED ISSUE #8767: [Issue] TopoNaming Phase 1 (card 3 of 4) - create MappedElement class
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Forums discussion

_No response_

### Version

0.20 (Release)

### Full version info

```shell
[code]
OS: Pop!_OS 22.04 LTS (pop:GNOME/pop)
Word size of FreeCAD: 64-bit
Version: 0.20.14555 (Git shallow) Snap 222
Build type: Debug
Branch: master
Hash: d4feb6578398d51eaf56cd794e8d3e2e000bea64
Python 3.10.6, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.5.1
Locale: English/United States (en_US)
[/code]
```

### Subproject(s) affected?

None

### Issue description

> The toponaming algorithm would benefit from a type to easily store and manipulate element naming as a whole.

## description

Create teh MappedElement.cpp class that is used to save the name parts of an element. Steps 1 and 2 must be complete first.

For context, see issue https://github.com/FreeCAD/FreeCAD/issues/8432

## notes

It contains the current IndexedName and the generated MappedName.

most of the work has been done already in the branch at `src/App/MappedElement.h`.

The comparator `ElementNameComp` and the classes `HistoryItem` and `MappedChildElements` will be added elsewhere in phase 2 when ElementMap code is built out.

mostly, just extract this into a separate file and add good unit tests to ensure that these core functions are very reliable.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by chennes**
Closes #8767.

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974">pipeline status</a> for feature branch [PR_8974](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974). Pipeline [813722402](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/813722402) was triggered at [84694cc](https://github.com/FreeCAD/FreeCAD/pull/8974/commits/84694cc8636a10acc1cbfa9e99035fa638ad992e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974">pipeline status</a> for feature branch [PR_8974](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974). Pipeline [813772904](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/813772904) was triggered at [f762ce2](https://github.com/FreeCAD/FreeCAD/pull/8974/commits/f762ce2a70c9d01d7bbb90fc5be04bd2e2e818f9). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: tests/src/App/MappedElement.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+{
+    // Index is compared first, then mappedName
+
+    // Arrange
+    std::vector<Data::IndexedName> indexedNames {
+        Data::IndexedName {"EDGE1"},
+        Data::IndexedName {"EDGE2"},
+        Data::IndexedName {"EDGE3"},
+    };
+    std::vector<Data::MappedName> mappedNames {
```
Comment (AjinkyaDahale) [inline tests/src/App/MappedElement.cpp:123]: Does this policy stay for faces and vertices? Is there a possibility they may need to be compared against each other?

File: tests/src/App/MappedElement.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    for (const auto& indexedName : indexedNames) {
+        for (const auto& mappedName : mappedNames) {
+            mappedElements.emplace_back(indexedName, mappedName);
+        }
+    }
+
+    // Act & Assert
+    for (size_t i = 0; i < mappedElements.size() - 1; ++i) {
+        EXPECT_TRUE(mappedElements[i] < mappedElements[i + 1]);
+    }
```
Comment (AjinkyaDahale) [inline tests/src/App/MappedElement.cpp:139]: Looks like overkill. Also, do we need to confirm equal mapped elements return false?

File: tests/src/App/MappedElement.cpp (hunk 1 above)
Comment (AjinkyaDahale) [reply tests/src/App/MappedElement.cpp:123]: Also, what about lowercase or other characters?

Comment (chennes) [reply tests/src/App/MappedElement.cpp:123]: I didn't want to be testing the operators of `IndexedName` and `MappedName`, which are well-tested and known to work, but simply to verify that `mappedName` uses them in the right order.

File: tests/src/App/MappedElement.cpp (hunk 2 above)
Comment (chennes) [reply tests/src/App/MappedElement.cpp:139]: Yeah, I could certainly simplify this to just a couple of predefined comparisons, and including the false comparison when equal is a good idea. I'll rewrite it.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974">pipeline status</a> for feature branch [PR_8974](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974). Pipeline [815870929](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/815870929) was triggered at [2a4e1af](https://github.com/FreeCAD/FreeCAD/pull/8974/commits/2a4e1aff401c36a4a14d7b873ac5c30180923b57). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974">pipeline status</a> for feature branch [PR_8974](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974). Pipeline [816373152](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/816373152) was triggered at [d2c5dce](https://github.com/FreeCAD/FreeCAD/pull/8974/commits/d2c5dce79efdcdffbd65ee790b64b7fb69316bad). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: tests/src/App/MappedElement.cpp (hunk 2 above)
Comment (AjinkyaDahale) [reply tests/src/App/MappedElement.cpp:138]: Also, false for `mappedElements[i + 1] < mappedElements[i]`.

Comment (AjinkyaDahale) [reply tests/src/App/MappedElement.cpp:138]: Sorry, can't see it. Only the old 9 element (3*3) vector.

Comment (chennes) [reply tests/src/App/MappedElement.cpp]: My bad -- I was fighting with some Windows compilation issues and overwrote my change 🤦 . Fixed up now, I swear I was trying to take your advice!! 😁

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974">pipeline status</a> for feature branch [PR_8974](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974). Pipeline [817087286](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/817087286) was triggered at [b909bc9](https://github.com/FreeCAD/FreeCAD/pull/8974/commits/b909bc9c59678d0620c78220fe9b7fc241c1c2df). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974">pipeline status</a> for feature branch [PR_8974](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8974). Pipeline [817092729](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/817092729) was triggered at [8b74cd3](https://github.com/FreeCAD/FreeCAD/pull/8974/commits/8b74cd3255886d0d9c637d6b3f1668bc167ac36c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: tests/src/App/MappedElement.cpp (hunk 2 above)
Comment (AjinkyaDahale) [reply tests/src/App/MappedElement.cpp]: That's alright. I was just getting a bunch of force push notifications and kept getting confused of what you changed.

File: tests/src/App/MappedElement.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+TEST_F(MappedElementTest, lessThanOperator)
+{
+    // Index is compared first, then mappedName
+
+    // Arrange
+    auto mappedElement1A = givenMappedElement("EDGE1", "A");
+    auto mappedElement1B = givenMappedElement("EDGE1", "B");
+    auto mappedElement2A = givenMappedElement("EDGE2", "A");
+    auto mappedElement2B = givenMappedElement("EDGE2", "B");
+    auto mappedElement2BDuplicate = givenMappedElement("EDGE2", "B");
```
Comment (AjinkyaDahale) [inline tests/src/App/MappedElement.cpp:121]: Some more questions which may be better in indexedname and mappedname but they are merged: what about, say "edge2" vs "edge12". Does it matter which is first in the order?

Comment (chennes) [reply tests/src/App/MappedElement.cpp:121]: They are sorted in numerical order (e.g. Edge2 comes before Edge12). I'm not sure if there is a test in IndexedName that verifies that, however.

Comment (AjinkyaDahale) [reply tests/src/App/MappedElement.cpp:121]: We may want to look in and add that. As such the test here is probably less important than there (the code is already as complex the test) IMO.

Comment (chennes) [reply tests/src/App/MappedElement.cpp:121]: https://github.com/FreeCAD/FreeCAD/pull/9048

**Pull request merged**